### PR TITLE
SMTP: STARTTLS failed (code: 220, response: 2.0.0 Ready to start TLS)

### PR DESCRIPTION
**Pull request by schengawegga**
SMTP: STARTTLS failed (code: 220, response: 2.0.0 Ready to start TLS) #68

### CODE REVIEW

**Review by jparise — APPROVED** [review]
Code looks good!

**Comment by schengawegga** [comment]
@jparise could you do another code-review, please?

File: Net/SMTP.php [hunk 1]
```diff
@@ -623,23 +623,43 @@ public function starttls()
...
                 }
-                if (PEAR::isError($result = $this->socket->enableCrypto(true, $crypto_method))) {
+
+                $result = 0;
+                while ($result == 0) {
+                    $result = $this->socket->enableCrypto(true, $crypto_method);
+                    if ($this->socket->isBlocking() !== true) {
+                        usleep(1);
+                    }
+                }
```
Comment (jparise) [inline Net/SMTP.php:648]: This _could_ loop indefinitely. Let's limit the number of attempts.

Comment (jparise) [reply Net/SMTP.php:648]: This could also be expressed as a `do .. while` loop. It always makes at least one iteratin.

Comment (schengawegga) [reply Net/SMTP.php]: @jparise thank you. good response. i pushed another commit with limiting the number of loop attempts.

File: Net/SMTP.php [hunk 2]
```diff
@@ -623,23 +623,47 @@ public function starttls()
...
-                if (PEAR::isError($result = $this->socket->enableCrypto(true, $crypto_method))) {
+
+                $do_count = 0;
+                do {
+                    $result = $this->socket->enableCrypto(true, $crypto_method);
+                    if ($this->socket->isBlocking() !== true) {
+                        usleep(1);
+                    }
+                    $do_count++;
+                } while ($result == 0 && $do_count < 100);
```
Comment (jparise) [inline Net/SMTP.php:649]: Now that I look at it, perhaps a little clearer using a `for` loop:

```suggestion
                for ($attempts = 1; $attempts < 100; $attempts++) {
                    $result = $this->socket->enableCrypto(true, $crypto_method);
                    if ($this->socket->isBlocking() !== true) {
                        usleep(1);
                    }                   
                    if ($result == 0) {
                        break;
                    }
                }

```

Comment (schengawegga) [reply Net/SMTP.php]: @jparise many roads lead to rome ;-)
But i like your suggestions, so i has committed it.
Could you do another code-review, please?

**Comment by schengawegga** [comment]
@jparise can you do another review, please?

**Comment by schengawegga** [comment]
@Neustradamus Did you received my mail?

File: Net/SMTP.php [hunk 3]
```diff
@@ -623,23 +623,69 @@ public function starttls()
...
+
+                    if (defined('STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT')) {
+                        $crypto_method |= @STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT;
+                    }
+
+                    if (defined('STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT')) {
+                        $crypto_method |= @STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT;
+                    }
+
+                    //New Crypto-Method since PHP7.4
```
Comment (jparise) [inline Net/SMTP.php:636]: I'm not sure this comment is necessary nearly all of the `STREAM_CRYPTO_METHOD_*` constants are conditional.

File: Net/SMTP.php [hunk 4]
```diff
@@ -623,23 +623,69 @@ public function starttls()
...
+                    if (defined('STREAM_CRYPTO_METHOD_TLSv1_3_CLIENT')) {
+                        $crypto_method |= @STREAM_CRYPTO_METHOD_TLSv1_3_CLIENT;
+                    }
                 }
-                if (PEAR::isError($result = $this->socket->enableCrypto(true, $crypto_method))) {
+
+                for ($attempts = 1; $attempts < 100; $attempts++) {
```
Comment (jparise) [inline Net/SMTP.php:642]: `100` is somewhat arbitrary. It means that will take at most ~100ms. Is that the upper time bound you had in mind?

Alternatively, we could set `$attempts` to something like `10` and use `usleep($attempts);` to spread out fewer attempts over a similar amount of time.

File: Net/SMTP.php [hunk 5]
```diff
@@ -623,23 +623,69 @@ public function starttls()
...
+                    $result = $this->socket->enableCrypto(true, $crypto_method);
```
Comment (jparise) [inline Net/SMTP.php:643]: Do we need to check for a `PEAR::isError($result)` result from this method call?

File: Net/SMTP.php [hunk 6]
```diff
@@ -623,23 +623,69 @@ public function starttls()
...
+
+                    return PEAR::raiseError('STARTTLS failed ' .
+                        '{enableCrypto: ' . var_export($result, true) . '; ' .
+                        'crypto_method: ' .
+                        array_search($crypto_method, $crypto_types_arr) .
+                        ' (' . var_export($crypto_method, true) . '); ' .
+                        'attempts: ' . $attempts . '; ' .
+                        array_search(error_get_last()['type'], $error_types_arr) .
+                        ' (' . error_get_last()['type'] . '): ' .
+                        error_get_last()['message'] . '}');
```
Comment (jparise) [inline Net/SMTP.php:679]: I think this would be more readable using [`sprintf()`](https://www.php.net/sprintf).

File: Net/SMTP.php (hunk 3 above)
Comment (schengawegga) [reply Net/SMTP.php:636]: You´re right. I will remove them in a new commit soon.

File: Net/SMTP.php (hunk 4 above)
Comment (schengawegga) [reply Net/SMTP.php:642]: That is a good suggestion. I will change this into `15` attempts, wich is equal to something like 120ms at maximum. I will commit that soon, too.

File: Net/SMTP.php (hunk 5 above)
Comment (schengawegga) [reply Net/SMTP.php:643]: There is a check for `PEAR::isError($result)` at line 651. Do you think a sooner check is necessary?

File: Net/SMTP.php (hunk 6 above)
Comment (schengawegga) [reply Net/SMTP.php:679]: I will check that and give response soon.

File: Net/SMTP.php (hunk 5 above)
Comment (jparise) [reply Net/SMTP.php:643]: Yes, but it is also possible for these inner (additional) checks to fail, too. I don't know if that should be considered a fatal error or now, though.

File: Net/SMTP.php (hunk 6 above)
Comment (schengawegga) [reply Net/SMTP.php]: @jparise I changed a few things to get better readability and backwards-compatibility to PHP5.4. Can you check my changes, please?

File: Net/SMTP.php [hunk 7]
```diff
@@ -1475,4 +1529,27 @@ public function identifySender()
...
+     * @param string $filter The string to filter
+     * @return array Filtered constants array.
+     *
+     * @since 1.10.2
+     */
+    private function getDefinedConstantsKeyFilter($filter) {
+        $constants_filtered = array();
+        $filter_length = strlen($filter);
+        $constants = get_defined_constants();
+        foreach($constants As $key=>$value){
```
Comment (jparise) [inline Net/SMTP.php:1548]:
```suggestion
        foreach ($constants as $key=>$value){
```

File: Net/SMTP.php [hunk 8]
```diff
@@ -1475,4 +1529,27 @@ public function identifySender()
...
+            if(substr($key,0,$filter_length) == $filter) {
```
Comment (jparise) [inline Net/SMTP.php:1549]:
```suggestion
            if (substr($key, 0, $filter_length) == $filter) {
```

File: Net/SMTP.php [hunk 9]
```diff
@@ -1475,4 +1529,27 @@ public function identifySender()
...
+    /**
+     * Backwards-compatibility method.
+     * array_filter alternative in PHP5.4 for using
+     * key filter because array_filter mode parameter
+     * is only available since PHP5.6.
+     *
+     * @param string $filter The string to filter
+     * @return array Filtered constants array.
+     *
+     * @since 1.10.2
```
Comment (jparise) [inline Net/SMTP.php:1542]: We don't need `@since` here because it's not part of the public API.

**Review by jparise — APPROVED** [review]
Just some minor reformatting

File: Net/SMTP.php (hunk 9 above)
Comment (schengawegga) [reply Net/SMTP.php]: ah i see. thanks. i deleted @since. can you have a look again?

**Review by jparise — APPROVED** [review]

**Pull request merged**